shpc/main/registry/remote.py

```diff
@@ -91,44 +74,69 @@ def override_exists(self, tag):
 
 
 class VersionControl(Provider):
-    def __init__(self, *args, **kwargs):
-        self.tag = kwargs.get("tag")
+
+    # The URL is substituted with user, repo
+    library_url_schemes = {
+        "github.com": "https://%s.github.io/%s/library.json",
+        "gitlab.com": "https://%s.gitlab.io/%s/library.json",
+    }
+
+    # The URL is substituted with repo, branch, module_name
+    web_container_url_schemes = {
+        "github.com": "https://github.com/%s/blob/%s/%s/container.yaml",
+        "gitlab.com": "https://gitlab.com/%s/-/blob/%s/%s/container.yaml",
+    }
+
+    # The URL is substituted with repo, branch, module_name
+    raw_container_url_schemes = {
+        "github.com": "https://raw.githubusercontent.com/%s/%s/%s/container.yaml",
+        "gitlab.com": "https://gitlab.com/%s/-/raw/%s/%s/container.yaml",
+    }
+
+    def __init__(self, source, tag=None, subdir=None):
+        if "://" not in source:
+            if os.path.exists(source):
+                raise ValueError(
+                    "VersionControl registry must be a remote path, not a local one."
+                )
+            # Normalise the URL
+            # Heuristics: if there is a @, it's probably ssh
+            if "@" in self.url:
+                self.url = "ssh://" + source
+            else:
+                self.url = "https://" + source
+        else:
+            self.url = source
+
+        self.is_cloned = False
+
+        self.tag = tag
 
         # Cache of remote container metadata
         self._cache = {}
 
         # E.g., subdirectory with registry files
-        self.subdir = kwargs.get("subdir")
-        super().__init__(*args, **kwargs)
-        self._url = self.source
+        self.subdir = subdir
 
     @classmethod
     def matches(cls, source):
-        return cls.provider_name in source and source.startswith("http")
+        return ("://" in source) or not os.path.exists(source)
```

Comment:
Any reason for the extra parens? they shouldn't be needed here - it's one or with two checks, so there is only one option for the order.

Reply:
That's my style :) I add parentheses to help my brain identify the logical blocks, even if they're not needed due to the precedence of the operators. I actually had this one as `return (("://" in source) or not os.path.exists(source))` initially but remove the outer set of parentheses because of `black` ;)

Reply:
Changed